File: TopInterview150/Hashmap/WordPattern/solution.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        ps = {}
        sp = {}
        words = s.split()
        if len(words) != len(pattern):
            return False
        for i in range(len(pattern)):
            word = words[i]
            char = pattern[i]
            if char not in ps:
                ps[char] = word
            elif ps[char] != word:
                return False
        for i in range(len(pattern)):
            word = words[i]
            char = pattern[i]
            if word not in sp:
                sp[word] = char
            elif sp[word] != char:
                return False
        return True
    """interestingly enough, keeping the two loops separate here runs at 13ms, but combining them into one runs at 19ms, even though it's more efficient generally"""

File: TopInterview150/Hashmap/WordPattern/solution.cpp
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        unordered_map<char, string> ps;
        unordered_map<string, char> sp;
        vector<string> words;
        istringstream stream(s);
        string word;

        //create the vector of words
        while (stream >> word)
        {
            words.push_back(word);
        }

        if (pattern.length() != words.size())
            return false;

        for (int i = 0; i < pattern.length(); i++)
        {
            word = words[i];
            char cha = pattern[i];
            if (ps.find(cha) == ps.end())
            {
                ps[cha] = word;
            }
            else if (ps[cha] != word)
            {
                return false;
            }
            if (sp.find(word) == sp.end())
            {
                sp[word] = cha;
            }
            else if (sp[word] != cha)
            {
                return false;
            }
        }
        return true;
    }
};